libs/vm/include/vm/variant.hpp

```diff
@@ -36,208 +36,211 @@ union Primitive
   {
     f32 = value;
   }
 
-  void Set(double value)
+  constexpr void Set(double value) noexcept
   {
     f64 = value;
   }
 
-  void Set(fixed_point::fp32_t value)
+  constexpr void Set(fixed_point::fp32_t value) noexcept
   {
     i32 = value.Data();
   }
 
-  void Set(fixed_point::fp64_t value)
+  constexpr void Set(fixed_point::fp64_t value) noexcept
   {
     i64 = value.Data();
   }
 };
 
+template <>
+constexpr auto Primitive::Get<bool>() const noexcept
+{
+  return bool(ui8);
+}
+
+template <>
+constexpr auto Primitive::Get<int8_t>() const noexcept
+{
+  return i8;
+}
+
+template <>
+constexpr auto Primitive::Get<uint8_t>() const noexcept
+{
+  return ui8;
+}
+
+template <>
+constexpr auto Primitive::Get<int16_t>() const noexcept
+{
+  return i16;
+}
+
+template <>
+constexpr auto Primitive::Get<uint16_t>() const noexcept
+{
+  return ui16;
+}
+
+template <>
+constexpr auto Primitive::Get<int32_t>() const noexcept
+{
+  return i32;
+}
+
+template <>
+constexpr auto Primitive::Get<uint32_t>() const noexcept
+{
+  return ui32;
+}
+
+template <>
+constexpr auto Primitive::Get<int64_t>() const noexcept
+{
+  return i64;
+}
+
+template <>
+constexpr auto Primitive::Get<uint64_t>() const noexcept
+{
+  return ui64;
+}
+
+template <>
+constexpr auto Primitive::Get<float>() const noexcept
+{
+  return f32;
+}
+
+template <>
+constexpr auto Primitive::Get<double>() const noexcept
+{
+  return f64;
+}
+
+template <>
+constexpr auto Primitive::Get<fixed_point::fp32_t>() const noexcept
+{
+  return fixed_point::fp32_t::FromBase(i32);
+}
+
+template <>
+constexpr auto Primitive::Get<fixed_point::fp64_t>() const noexcept
+{
+  return fixed_point::fp64_t::FromBase(i64);
+}
+
 struct Variant
 {
   union
   {
     Primitive   primitive;
     Ptr<Object> object;
   };
-  TypeId type_id;
+  TypeId type_id = TypeIds::Unknown;
 
-  Variant()
+  Variant() noexcept
   {
     Construct();
   }
 
-  Variant(Variant const &other)
+  Variant(Variant const &other) noexcept
   {
     Construct(other);
   }
 
-  Variant(Variant &&other)
+  Variant(Variant &&other) noexcept
   {
     Construct(std::move(other));
   }
 
-  template <typename T, typename std::enable_if_t<IsPrimitive<T>::value> * = nullptr>
-  Variant(T other, TypeId other_type_id)
+  template <typename T, std::enable_if_t<IsPrimitive<T>::value> * = nullptr>
+  Variant(T other, TypeId other_type_id) noexcept
   {
     Construct(other, other_type_id);
   }
 
   template <typename T, typename std::enable_if_t<IsPtr<T>::value> * = nullptr>
-  Variant(T const &other, TypeId other_type_id)
+  Variant(T const &other, TypeId other_type_id) noexcept
   {
     Construct(other, other_type_id);
   }
 
   template <typename T, typename std::enable_if_t<IsPtr<T>::value> * = nullptr>
-  Variant(T &&other, TypeId other_type_id)
+  Variant(T &&other, TypeId other_type_id) noexcept
   {
-    Construct(std::forward<T>(other), other_type_id);
+    Construct(std::move(other), other_type_id);
```

Comment:
Why not `std::forward`?

Reply:
Currently, `IsPtr` is only true for `Ptr<T>` so this overload is only considered when `T` is `Ptr<U>` for some `U`, so `other` is an rvalue reference, and this wording is shorter.
I'm not sure, if `IsPtr` is ever to be true on references, then this line should be left intact. Should it?

Reply:
👍 - nope it shouldn't be - do you think that it is worth some kind of static assert here to catch that?

Reply:
@ejfitzgerald  has a point - as things stand now `move` is appropriate here, but only under assumptions which are implicit in the code and which would be easy to violate by making reasonable changes to client code. Given that `forward` works, has no performance penalty, and requires no assertions, it is preferable to `move` here.